src/network/handler.ml

```diff
@@ -0,0 +1,80 @@
+open Deku_concepts
+open Deku_consensus
+open Deku_chain
+open Deku_stdlib
+
+module type HANDLER = sig
+  type input
+  type output
+
+  val handler : input -> Chain.t -> (output, Internal_error.t) result
+  val output_to_response : output -> Piaf.Response.t
+end
+
+module Make_handler (Handler : HANDLER) : sig
+  val handle : path:string -> chain:Chain.t -> Handler.input -> Piaf.Response.t
+end = struct
+  open Piaf
+
+  let error ~message =
+    let open Internal_error in
+    let status = Status.of_code message.http_code in
+    let body = Internal_error.yojson_of_t message |> Yojson.Safe.to_string in
+    Response.of_string ~body ~headers:Endpoint.headers status
+
+  let handle ~path ~chain input =
+    let _ = path in
+    Log.info "handle request in %s" path;
+    (* Log.info "handle request on path %s" path; *)
+    let result = Handler.handler input chain in
+    match result with
+    | Error err ->
+        Log.error "error in path %s: %s" path (Internal_error.to_string err);
+        error ~message:err
+    | Ok output -> Handler.output_to_response output
+end
+
+module Header = struct
+  open Piaf
+
+  let allow_methods methods headers = Headers.add headers "Allow" methods
+
+  let allow_origin origin headers =
+    Headers.add headers "Access-Control-Allow-Origin" origin
+
+  let allow_headers headers =
+    Headers.add headers "Access-Control-Allow-Headers" "*"
+
+  let json_content headers =
+    Headers.add headers "Content-Type" "application/json"
+
+  let no_cache headers =
+    Headers.add headers "Cache-Control" "no-store, no-cache, max-age=0"
+
+  let content_length body headers =
+    Headers.add headers "Content-Length" (String.length body |> Int.to_string)
+end
+
+module Get_block_by_level = Make_handler (struct
+  type input = Level.t
+  type output = Block.t
+
+  let handler level chain =
+    (* TODO: get the block from the chain at the current level *)
+    let _ = level in
+    let _ = chain in
+    Genesis.block
+    |> Result.ok (* Should be replaced by the block at the given level *)
+
+  let output_to_response output =
+    let open Piaf in
+    (* TODO: Convert output to a readable output *)
+    let _ = output in
+    let body = "block as json" in
+    let headers =
+      Headers.empty |> Header.allow_methods "GET" |> Header.allow_origin "*"
```

Comment:
Yes I agree, I have mixed feeling about this.
Do you think it's better to allow all methods ? It seems strange to do so, but let's say the "allow method" header only contains "GET", if you do a POST, PUT (or what you want) on this route, the client will never have the error message "METHOD_NOT_ALLOWED", instead he will have "Cors Origin blablabla"
If you think it's better to only allow possible methods, i'll add the sumtype (maybe the one from piaf)

Reply:
You should/must only allow the correct methods per endpoint. An endpoint cannot be a `GET` and a `PUT` at the same time, since the `PUT` requires a body, which is not the case for the `GET` one.

Reply:
Ok, but I won't be able to return 405 and a custom message